### ISSUE TITLE
[feature request]  `-lco RFC8142=NO` for GeoJSONseq to allow other CRS

### ISSUE THREAD

**Issue by tobwen** [OSGeo/gdal#8659]
## featury request
I would like to see CRSs other than EPSG:4326 supported.

## other issues, RFC
I am well aware of https://github.com/OSGeo/gdal/issues/6859 and RFC 8142, but even now this directive is not strictly enforced (e.g. with RS, which is only set with a switch or with the extension `geojsons`).

## accessibility
GeoJSON is one of the few plain text formats, alongside XML-based formats and CSV, in which attributes and geometries can be viewed and changed with any software.  With regard to OpenData archives, such plain text formats may also be of interest in the future. This means that no software is required and the data can be interpreted immediately - even by humans. 

## raw data
The background to my wish for a different CRS is that the official, measured coordinates, e.g. in metres with three decimal places, are often stored in the national survey. When converting to EPSG:4326, this information could be lost - also because most people do not realise how the epochs (etc.) are handled. By allowing other CRS, those measured data can be entered, stored/archived in raw and original format.

## defaults
To ensure that existing scripts are not affected, you could create a new switch, analogous to the normal GeoJSON driver simple `-lco RFC8142=NO`. If this flag is set, `-t_srs` could take place without transformation to EPSG:4326 (analogue to the normal GeoJSON driver).

**Comment by jratike80** [comment]
Alternatively there could be a JSONFGSeq driver based on https://gdal.org/drivers/vector/jsonfg.html. I believe that some people have strong disgust for crs in anything that is called GeoJSON.

**Comment by tobwen** [comment]
Of course, I agree with that. I don't want to take away or restrict the standard, but the function should be extended.

GDAL is a application that is used everywhere: from land surveying over geodetics and cartography to professional science and community or hobby projects. There are just use cases where you have to go beyond RFCs.

For me, GeoJSON has some advantages GML, but all CRSs are permitted there. It is therefore a common exchange format. For me, it isn't clear why GeoJSON was artificially restricted in this way. A convention could have been chosen: If no CRS is specified, then 4326. But all others are banned? That is narrow-minded.

**Comment by jratike80** [comment]
I think that we can limit the discussion to GeoJSONSeq driver. As you indirectly mentioned in your question, the GeoJSON driver does support crs by the 2008  version of the specification. It is even the default behavior for the driver and as far as I know lots of GDAL users utilize that possibility when they need to use other crs than CRS84 without bad conscience.

The problem for users like us at the National Land Survey of Finland is that there is no binding standard about how to support crs in GeoJSON. Therefore we will change into JSONFG once the standard is accepted in OGC and once enough many client libraries support JSONFG. We'll do that even for us the only real advantage in JSONFG is the standardized crs support. At least until that switch our official format is GML because our data are natively in EPSG:3067.

Out of curiosity, why GeoJSONSeq is important for you over GeoJSON? You mention archiving but what is the advantage in that use?

**Comment by tobwen** [comment]
> Out of curiosity, why GeoJSONSeq is important for you over GeoJSON? You mention archiving but what is the advantage in that use?

It was a pro-argument in favour of a "truly" open file format (i.e. in the sense that you can access the data by the simplest means). In a few decades, users might ask themselves what SHP or FGB mean - similar to the situation with E00 today.

I've got to deal with lots of data every day and, beside PostGIS and Python of course, I often use linux shell tools to process single line GeoJSONseq data. I can process huge GeoJSON files over 64 cores within seconds, while GPKG, FGB & Co. are locked down to a few instances or even just one. For some data, I even work with CSV data and geometries in WKB. But that's going offtopic.

> Therefore we will change into JSONFG once the standard is accepted in OGC and once enough many client libraries support JSONFG.

So this means to me, it's better to spend some weekends in understanding the code and make the mentioned `RFC8142=NO `switch work in my own fork :-)

**Comment by rouault** [comment]
(side note/minor point: it would be odd to name the option RFC8142=NO, because RFC8142 is the RFC that defines GeoJSONSeq)
A technical issue with allowing other CRS in GeoJSONSeq is that there would be no way of storing it in the file... Unless you store it in each feature.
And more fundamentally that would impact the interoperability story of GeoJSONSeq
If you use GeoJSON 2008 as output by GDAL, given the way it format the data, you can process it line by line like you would do with GeoJSONSeq once you've skipped the first few lines

CC @sgillies

**Comment by tobwen** [comment]
> If you use GeoJSON 2008 as output by GDAL, given the way it format the data, you can process it line by line like you would do with GeoJSONSeq once you've skipped the first few lines

Yeah, that's what I do right now.

Okay, I'm closing this, since it's a wontfix/wontdo. Thanks for your time, guys.

**Comment by sgillies** [comment]
> (side note/minor point: it would be odd to name the option RFC8142=NO, because RFC8142 is the RFC that defines GeoJSONSeq)
A technical issue with allowing other CRS in GeoJSONSeq is that there would be no way of storing it in the file... Unless you store it in each feature.
And more fundamentally that would impact the interoperability story of GeoJSONSeq

Yes, this is a major sticking point.

> If you use GeoJSON 2008 as output by GDAL, given the way it format the data, you can process it line by line like you would do with GeoJSONSeq once you've skipped the first few lines

This is an undocumented feature of GDAL's GeoJSON writer, though. Can people really depend on it?

For the 1000th time, let me remind all that GeoJSON isn't designed to be used by national mapping agencies. It's designed to be used alongside HTML and HTTP for web applications and infrastructure with a global scope and on systems with no EPSG database and no PROJ installation. Imagine if HTML didn't have a single standard set of tags with English names and ASCII encoding, but different countries or organizations used different names and encoding. Imagine if a successfully HTTP request didn't return `200 OK` but some nation-specific number and code. Which required a proprietary database like EPSG and complex, hefty software and data like PROJ and its grids for translation and interpretation. That's what non-standard GeoJSON would be like.

That said, I *do* use streams of projected GeoJSON-like features all the time in my work. The `fio cat` command from the Fiona Python package streams "GeoJSON" out from datasets in the native coordinate system and can reproject as well. For example:

```
% fio cat zip+https://s3.amazonaws.com/fiona-testing/coutwildrnp.zip --dst-crs EPSG:32612 | head -n1 | jq
{
  "bbox": [
    435895.1516480881,
    4621697.96707111,
    447644.7208376831,
    4649573.068823434
  ],
  "geometry": {
    "coordinates": [
      [
        [
          439101.39600850403,
          4649492.9996704245
        ],
        ...,
        [
          439101.39600850403,
          4649492.9996704245
        ]
      ]
    ],
    "type": "Polygon"
  },
  "id": "0",
  "properties": {
    "AGBUR": "FS",
    "AREA": 0.0179264,
    "FEATURE1": "Wilderness",
    "FEATURE2": null,
    "NAME": "Mount Naomi Wilderness",
    "PERIMETER": 1.22107,
    "STATE": "UT",
    "STATE_FIPS": "49",
    "URL": "http://www.wilderness.net/index.cfm?fuse=NWPS&sec=wildView&wname=Mount%20Naomi",
    "WILDRNP020": 332
  },
  "type": "Feature"
}
```

Note that if the output is streamed into another program like `fio load`, the coordinate reference system must be understood by convention or communicated [out of band](https://en.wikipedia.org/wiki/Out-of-band). I don't find this a huge limitation, in practice.

**Comment by jratike80** [comment]
> For the 1000th time, let me remind all that GeoJSON isn't designed to be used by national mapping agencies.

I would put that "GeoJSON isn't designed to be used for delivering data in other crs than CRS84 without out-of-band communication".

I do not know other mapping agencies than NLS well, but was GeoJSON designed for us or not, we do use GeoJSON a lot. Our users get GeoJSON data through our APIs and as downloadable files. Both CRS84 and some other systems are supported. There is a minor issue when the users have saved something like this into a file on their computer and then they forget what was the "a prior arrangement" that they made with our API
`"geometry":{"type":"Polygon","coordinates":[[[29349626.39,7389491.765],[29349630.075,7389493.37],[29349631.803,7389489.402],[29349628.118,7389487.798],[29349626.39,7389491.765]]]}`

If I tell that the polygon is in Finland ("understood by convention") it is not super hard to find out the CRS (try!). For our clients we say that they can download GML instead or then they can take notes when they download GeoJSON. That is not a huge limitation, users learn soon to take notes.

I tried to think if the HTML example is good or not but I am an agronomist and not an IT person so I can't really judge. For me GeoJSON feels like a document and HTTP can deliver documents which are not in ASCII or English. It is even prepared to discuss with clients about that https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Language. However, people making web mapping applications obviously, and probably for a good reason, feel that GeoJSON is more than a document.

In conclusion, NLS and I take that the GeoJSON format is fixed.

**Comment by tobwen** [comment]
For archiving and community reasons, that's the code I'm using to produce GeoJSONseq with other CRS (`EPSG:25832` in this example) using the GeoJSON driver:

```sh
ogr2ogr /vsistdout/ -F GeoJSON dataset.gpkg layer -t_srs EPSG:25832 -lco COORDINATE_PRECISION=4 \
| awk 'NR > 4 && /^{/ { sub(/,$/, ""); print }' \
| pv -ls50M >/dev/null
```
I'm getting 191k lines per second, the overhead is negligible.
